src/modules/cxxpiper.cpp

```diff
@@ -0,0 +1,1223 @@
+    {
+	char *fn = NULL;
+	if (strlen(SoundIconFolder) == 0) return fn;
+	GString *filename = g_string_new(SoundIconFolder);
+	filename = g_string_append(filename, icon_name);
+	if (g_file_test(filename->str, G_FILE_TEST_EXISTS)) fn = filename->str;
+	/* if the file was found, the pointer *fn  points to the character data
+	   of the string filename. In this situation the string filename must be
+	   freed but its character data must be preserved.
+	   If the file is not found, the pointer *fn contains NULL. In this
+	   situation the string filename must be freed, including its character
+	   data.
+	*/
+	return g_string_free(filename, (fn == NULL));
+    }
+
+    void cxxpiper_stretch_and_copy(const int samplerate, const int channels, vector<int16_t>& audioBuffer, vector<int16_t>& sharedAudioBuffer)
+    {
+	// In rubberband stretcher units, 1.0 is no duration change.
+	// Map -100 .. 100 onto 0.0 .. 2.0 .
+	double ratio = (((double)msg_settings.rate) + 100.0) / 100.0;
+	// Pitch is converted to frequency internally, as per 'piper'
+	// command line utility.  The 'adjust' function uses the same
+	// conversion.  I think -10.0 to 10.0
+	// provides a usable range plus a bit of margin.
+	// .  Map -100 .. 100 onto -10.0 .. 10.0
+	// Also clamp input values otherwise we will crash!
+	double mS = ((double)msg_settings.pitch);
+	if (mS < -100.0) mS = -100.0;
+	if (mS > 100.0) mS = 100.0;
+	double pitchshift = mS / 10.0;
+	//  In rubberband stretcher units, 1.0 is no gain/attenuation.
+	// Map -100 .. 100 onto 0.0 .. 1.0
+	float gain = (((float)msg_settings.volume) + 100.0f) / 200.0f;
+	(void)adjust(samplerate, channels, ratio, pitchshift, gain, audioBuffer, sharedAudioBuffer);
+    }
+
+    static void cxxpiper_handle_text(const char *data)
+    {
+	DBG("Input data before strip XML: %s", data);
+	cmdInp = (char *) module_strip_ssml(data);
+	DBG("Input after strip XML: %s", cmdInp);
+	mutex mutAudio;
+	condition_variable cvAudio;
+	bool audioReady = false;
+	cbCnt = 0;
+	cbTot = 0;
+	prCnt = 0;
+	prTot = 0;
+	izCnt =0;
+	izTot =0;
+	ezCnt =0;
+	ezTot =0;
+	runConfig.lengthScale = msg_settings.rate / 100.0;
+	vector<int16_t> audioBuffer;
+	vector<int16_t> sharedAudioBuffer;
+	piper::SynthesisResult result;
+	auto audioCallback = [&audioBuffer, &sharedAudioBuffer, &mutAudio,
+			      &cvAudio, &audioReady]()
+	{
+	    unique_lock lockAudio(mutAudio);
+#if 0
+	    copy(audioBuffer.begin(), audioBuffer.end(), back_inserter(sharedAudioBuffer));
+#else
+	    cxxpiper_stretch_and_copy(voice.synthesisConfig.sampleRate, voice.synthesisConfig.channels, audioBuffer, sharedAudioBuffer);
+#endif
+	    audioReady = true;
+	    ++cbCnt;
+	    cbTot += audioBuffer.size();	    //cerr << "CVaudio notivy 1" << endl;
+	    cvAudio.notify_one();
+	};
+	DBG("Sending begin event");
+	module_report_event_begin();
+	(void)cxxpiper::textToAudio(piperConfig, voice, cmdInp,
+				    audioBuffer, result, audioCallback);
```

Comment:
Does piper support pipelining? To start sending audio to the server before the whole audio is produced. That would be really important to get good reactivity (but it's fine not to have it for first inclusion)